aptos-move/framework/src/natives/cryptography/ristretto255_scalar.rs

```diff
@@ -0,0 +1,286 @@
+    _context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(_ty_args.is_empty());
+    debug_assert!(arguments.len() == 1);
+
+    let mut cost = gas_params.base_cost;
+
+    let bytes = pop_arg!(arguments, Vec<u8>);
+    if bytes.len() != 32 {
+        return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)]));
+    }
+
+    let bytes_slice = <[u8; 32]>::try_from(bytes).unwrap();
+
+    let s = Scalar::from_canonical_bytes(bytes_slice);
+    cost += gas_params.scalar_is_canonical_cost * NumArgs::one();
+
+    // TODO: Speed up this implementation using bit testing on 'bytes'?
+    Ok(NativeResult::ok(cost, smallvec![Value::bool(s.is_some())]))
+}
+
+pub(crate) fn native_scalar_invert(
+    gas_params: &GasParameters,
+    _context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(_ty_args.is_empty());
+    debug_assert!(arguments.len() == 1);
+
+    let mut cost = gas_params.base_cost;
+
+    let s = pop_scalar_from_bytes(&mut arguments)?;
+
+    // We'd like to ensure all Move Scalar types are canonical scalars reduced modulo \ell
+    debug_assert!(s.is_canonical());
+
+    // Invert and return
+    cost += gas_params.scalar_invert_cost * NumArgs::one();
+    Ok(NativeResult::ok(
+        cost,
+        smallvec![Value::vector_u8(s.invert().to_bytes().to_vec())],
+    ))
+}
+
+pub(crate) fn native_scalar_from_sha512(
+    gas_params: &GasParameters,
+    _context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(_ty_args.is_empty());
+    debug_assert!(arguments.len() == 1);
+
+    let mut cost = gas_params.base_cost;
+
+    let bytes = pop_arg!(arguments, Vec<u8>);
+
+    cost += gas_params.scalar_uniform_from_64_bytes_cost * NumArgs::one()
+        + gas_params.sha512_per_hash_cost * NumArgs::one()
+        + gas_params.sha512_per_byte_cost * NumBytes::new(bytes.len() as u64);
+    let s = Scalar::hash_from_bytes::<Sha512>(bytes.as_slice());
+
+    Ok(NativeResult::ok(
+        cost,
+        smallvec![Value::vector_u8(s.to_bytes().to_vec())],
+    ))
+}
+
+pub(crate) fn native_scalar_mul(
+    gas_params: &GasParameters,
+    _context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(_ty_args.is_empty());
+    debug_assert!(arguments.len() == 2);
+
+    let mut cost = gas_params.base_cost;
+
+    let b = pop_scalar_from_bytes(&mut arguments)?;
+    let a = pop_scalar_from_bytes(&mut arguments)?;
+
+    // We'd like to ensure all Move Scalar types are canonical scalars reduced modulo \ell
+    debug_assert!(a.is_canonical());
+    debug_assert!(b.is_canonical());
+
+    cost += gas_params.scalar_mul_cost * NumArgs::one();
+    let s = a.mul(b);
```

Comment:
why not just compute the gas all at once?

Reply:
Oh, in this case, I could. In other cases, I might fail before the expensive operations, so I don't.

I will fix appropriately.